### PR TITLE
Pure async RpcClient

### LINKED ISSUE #209: RpcClient needs to support async versions of RPC wrapper methods
For example, WalletApi.Transfer needs a WalletApi.TransferAsync counterpart

### PR DESCRIPTION
**Pull request by devhawk**
This PR plumbs removes all sync RpcClient calls and plumbs async thru end to end. 

If someone wants to block on an RpcClient call, they can use `.Result` (or `.GetAwaiter().GetResult()`) on the top level Task object.

~~Note, there are a few tests that are failing, so this PR isn't quite ready to merge yet. But I wanted to open it for review feedback since the core design is complete.~~ All the tests are passing now.

fixes #209

### CODE REVIEW

**Comment by ProDog** [comment]
Need wait https://github.com/neo-project/neo-modules/issues/311 fixed. 
Otherwise NetworkFee is inaccurate.

**Comment by devhawk** [comment]
> Need wait #311 fixed. Otherwise NetworkFee is inaccurate.

I'm fine with waiting, but RpcClient is currently building against Neo 3.0.0-preview3-00. I'm thinking it would make more sense to merge this PR as is and then update RpcClient to a more recent build of Neo to get the correct NetworkFee as a separate task.

**Review by ProDog — DISMISSED** [review]

**Review by shargon — COMMENTED** [review]
It's not recommended to add `Async` at the end of the method name?

File: src/RpcClient/TransactionManager.cs
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:195]: We should wait for https://github.com/neo-project/neo-modules/pull/338

**Comment by devhawk** [comment]
> It's not recommended to add `Async` at the end of the method name?

I didn't think that it was when you don't provide sync and async versions of methods. However, based on [official MSFT docs](https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/task-based-asynchronous-pattern-tap), it appears that I should: " Asynchronous methods in TAP include the Async suffix after the operation name for methods that return awaitable types, such as Task, Task<TResult>, ValueTask, and ValueTask<TResult>." Will fix

**Comment by devhawk** [comment]
Now that #338 is merged and #311 is fixed, I've updated this PR and we're ready to merge.

Note, I've added the fix for #343 to this PR. `TransactionManager` now supports an optional magic constructor parameter (defaults to ProtocolSettings.Default.Magic if unspecified) which is then passed to a recently added overload of `Neo.Wallets.Helper.Sign`

**Comment by shargon** [comment]
I will review it

File: src/RpcClient/ContractClient.cs
Comment (vncoelho) [inline src/RpcClient/ContractClient.cs:35]: This one is not `public async`?

Comment (devhawk) [reply src/RpcClient/ContractClient.cs:35]: no, because it doesn't use await. Since TestInvoke doesn't need to process the results of InvokeScriptAsync, it can simply return the Task<RpcInvokeResult> returned by InvokeScriptAsync

**Comment by devhawk** [comment]
@shargon, are we still waiting on anythign for this PR?

**Comment by shargon** [comment]
> @shargon, are we still waiting on anything for this PR?

No, we can review and merge now

File: src/RpcClient/TransactionManager.cs
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:43]: Add summary for `magic` ?

File: src/RpcClient/TransactionManager.cs
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:77]: This will not guarantee that the transaction will be with `ValidUntilBlock` and `SystemFee` set at the end of the method, will it? because the Tx it's public and can be used before `Sign`

File: src/RpcClient/TransactionManager.cs
Comment (shargon) [inline src/RpcClient/TransactionManager.cs:167]: if we call again `MakeTransaction` the queue will be empty?

**Review by shargon — COMMENTED** [review]
@cloud8little could you test these changes?

File: src/RpcClient/TransactionManager.cs
Comment (devhawk) [reply src/RpcClient/TransactionManager.cs:77]: This is true. I was trying to maintain the fluent API, but if this is a concern then MakeTransaction needs to be async. will fix

File: src/RpcClient/TransactionManager.cs
Comment (devhawk) [reply src/RpcClient/TransactionManager.cs:167]: If you call `MakeTransaction` again, it will queue two new items.

Based on this + the earlier comment about using the Tx before `ValidUntilBlock` and `SystemFee` are set, I'm going to redesign this API a bit.

**Comment by devhawk** [comment]
Note to self - need to add magic parameter to helper methods that create TxManager (CreateDeployContractTxAsync, CreateTransferTxAsync, CreateTransferTxAsync)

**Comment by devhawk** [comment]
Based on @shargon's observation that Tx is publicly accessible before `ValidUntilBlock` and `SystemFee`, I've redesgined the api so that MakeTransaction is async. I've moved MakeTransactionAsync to a new class TransactionManagerFactory to make it clear that a TransactionManager instance is only good for signing a single transaction. This approach allowed me to simplify TransactionManager and remove the QueueWork infrastructure. It's not as fluid as before - there are two async operations instead of one - but it ensures the incomplete Tx instance is never visible.

Usage:

``` csharp
var factory = new TransactionManagerFactory(rpcClient, magic);
var manager = await factory.MakeTransactionAsync(script, signers);
var tx = await manager
    .AddSignature(key)
    .SignAsync();
```

I've also provided a helper for scenarios when the user is only creating a single `TransactionManager`:

``` csharp
var manager = await TransactionManager.MakeTransactionAsync(rpcClient, script, signers, attributes, magic);
var tx = await manager
    .AddSignature(key)
    .SignAsync();
```

**Comment by shargon** [comment]
@superboyiii can your team test it?

**Comment by Tommo-L** [comment]
Waiting for some days, they are busy testing the DeFi projects.

File: src/RpcClient/RpcClient.cs
Comment (cloud8little) [inline src/RpcClient/RpcClient.cs:363]: invokescript also need sender as input. 
https://github.com/neo-project/neo-modules/blob/fed3f7a87d65d91bbe4e42bce94734cc76571d5c/src/RpcServer/RpcServer.SmartContract.cs#L117-L123

File: src/RpcClient/ContractClient.cs
Comment (cloud8little) [inline src/RpcClient/ContractClient.cs:67]: need sender as input.

File: src/RpcClient/Nep5API.cs
Comment (cloud8little) [inline src/RpcClient/Nep5API.cs:124]: need sender as input because invokescript rpc request need.

File: src/RpcClient/Nep5API.cs
Comment (cloud8little) [inline src/RpcClient/Nep5API.cs:150]: need sender also.

File: src/RpcClient/ContractClient.cs
Comment (devhawk) [reply src/RpcClient/ContractClient.cs:67]: Didn't realize RpcServer had changed w/o updating RpcClient. Will fix

File: src/RpcClient/RpcClient.cs
Comment (Tommo-L) [reply src/RpcClient/RpcClient.cs:392]: see https://github.com/neo-project/neo-modules/pull/368

**Comment by cloud8little** [comment]
@devhawk use the previous commit f3b06f5 and https://github.com/neo-project/neo-modules/pull/368, it works well for sendrawtransaction. I will continue testing nep5 methods and basic rpc commands.

**Comment by devhawk** [comment]
@cloud8little I reset my topic branch to f3b06f5c72fc7a21430ecd37f8feed96869843a5 and force pushed

**Comment by cloud8little** [comment]
All tests passed, including nep5APIs name/symbol/totalSupply and getblock*, getrawtransaction/sendrawtransaction; send*; transactionmanager, dump/import private key,etc.

File: src/RpcClient/RpcClient.cs
Comment (devhawk) [reply src/RpcClient/RpcClient.cs:265]: Can do, but it would be nice to get this PR merged and then make updates to master than keep having to adjust this PR to other changes. I'm going on vacation later this week, so I won't be able to keep adapting this PR in advance of preview 4

File: src/RpcClient/WalletAPI.cs
Comment (shargon) [inline src/RpcClient/WalletAPI.cs:120]: Maybe we can save the balance of, and transfer 0 tokens

Comment (shargon) [reply src/RpcClient/WalletAPI.cs:120]: https://github.com/neo-project/neo/blob/6886991ca67f2c2148429ba87bf7d782f8cb12c8/src/neo/SmartContract/Native/Tokens/NeoToken.cs#L49

Comment (devhawk) [reply src/RpcClient/WalletAPI.cs:120]: I think we should limit this PR to focus on making the existing RpcClient async. Suggestions like this can be made in later PRs and after we ship preview 4.

**Review by shargon — APPROVED** [review]

**Comment by superboyiii** [comment]
@erikzhang Shall you review as well? I will merge it one day later if no oppose since many PRs is waiting for this to be merged.

**Review by Tommo-L — APPROVED** [review]

**Pull request merged**
